Build failure with `template-haskell-2.17`, bound not revisable on hackage

Build with GHC 9.0 fails because of a change in `template-haskell` (see below).
A bound `template-haskell < 2.17` would be necessary for the versions already released on hackage.
Unfortunately, `template-haskell` is only a `build-depends` in conditionals in the cabal file and those cannot be revised.
(Due to this limitation, it is best practice to mention all dependencies in an unconditional `build-depends`, and add additional constraints on these dependencies in the conditionals.)
```
$ cabal install lenses

Build profile: -w ghc-9.0.1 -O1
In order, the following will be built (use -v for more details):
 - lenses-0.1.8 (lib:lenses) (requires build)
...
[2 of 2] Compiling Data.Lenses.Template ( src-26/Data/Lenses/Template.hs, dist/build/Data/Lenses/Template.o, dist/build/Data/Lenses/Template.dyn_o )

src-26/Data/Lenses/Template.hs:110:18: error:
    • Expecting one more argument to ‘TyVarBndr’
      Expected a type, but ‘TyVarBndr’ has kind ‘* -> *’
    • In the type signature:
        makeAccs :: [TyVarBndr] -> [VarStrictType] -> Q [Dec]
      In an equation for ‘nameDeriveLenses’:
...
```

--- a/lenses.cabal
+++ b/lenses.cabal
@@ -1,47 +1,62 @@
-Name:                lenses
-Version:             0.1.8
-Author:              Job Vranish
-Maintainer:          jvranish@gmail.com
-Homepage:            http://github.com/jvranish/Lenses/tree/master
-License:             BSD3
-License-file:        LICENSE
-category:            Data
-Synopsis:            Simple Functional Lenses
-Description:         A simple but powerful implementation of function lenses (aka functional
-                     references, accessors, etc..).
-                     They allow an easy way to manipulate fields of records
-Build-Type:          Simple
-Cabal-Version:       >= 1.6
+name:               lenses
+version:            0.1.8
+author:             Job Vranish
+maintainer:         jvranish@gmail.com
+homepage:           http://github.com/jvranish/Lenses/tree/master
+license:            BSD3
+license-file:       LICENSE
+category:           Data
+synopsis:           Simple Functional Lenses
+description:
+  A simple but powerful implementation of function lenses (aka functional
+  references, accessors, etc..).
+  They allow an easy way to manipulate fields of records
 
-Extra-Source-Files:
+build-type:         Simple
+cabal-version:      >=1.6
+extra-source-files:
   src-23/Data/Lenses/Template.hs
   src-24/Data/Lenses/Template.hs
   src-25/Data/Lenses/Template.hs
   src-26/Data/Lenses/Template.hs
+  src-27/Data/Lenses/Template.hs
 
 library
-  exposed-modules: Data.Lenses,
-                   Data.Lenses.Template
-  build-depends:  base >= 4 && < 6,
-                  mtl >= 1.1
-  If impl(ghc)
-     If impl (ghc >= 8.0)
-       HS-Source-Dirs: src-26
-       build-depends:  template-haskell >= 2.11.0.0
-     Else
-       If impl (ghc >= 7.10)
-         HS-Source-Dirs: src-25
-         build-depends:  template-haskell >= 2.10.0.0
-       Else
-          If impl(ghc >= 6.12)
-             Hs-Source-Dirs: src-24
-             Build-Depends:  template-haskell >= 2.4
-          Else
-             Hs-Source-Dirs: src-23
-             Build-Depends:  template-haskell >=2.2 && <2.4
-  hs-source-dirs: src/
-  ghc-options: -Wall
+  exposed-modules:
+    Data.Lenses
+    Data.Lenses.Template
+
+  build-depends:
+      base  >=4   && <6
+    , mtl   >=1.1
+
+  if impl(ghc)
+    if impl(ghc >=9.2)
+      hs-source-dirs: src-27
+      build-depends:  template-haskell >=2.17.0.0
+
+    else
+      if impl(ghc >=8.0)
+        hs-source-dirs: src-26
+        build-depends:  template-haskell >=2.11.0.0
+
+      else
+        if impl(ghc >=7.10)
+          hs-source-dirs: src-25
+          build-depends:  template-haskell >=2.10.0.0
+
+        else
+          if impl(ghc >=6.12)
+            hs-source-dirs: src-24
+            build-depends:  template-haskell >=2.4
+
+          else
+            hs-source-dirs: src-23
+            build-depends:  template-haskell >=2.2 && <2.4
+
+  hs-source-dirs:  src/
+  ghc-options:     -Wall
 
 source-repository head
-  type: git
+  type:     git
   location: git@github.com:jvranish/Lenses.git

--- a/src-27/Data/Lenses/Template.hs
+++ b/src-27/Data/Lenses/Template.hs
@@ -1,0 +1,144 @@
+{-# LANGUAGE TemplateHaskell,
+    CPP #-}
+
+{- |
+This module provides an automatic Template Haskell
+routine to scour data type definitions and generate
+lense objects for them automatically.
+
+It was copied almost verbatim (2 line change) from the wonderful Data.Accessors.Template module 
+made by Luke Palmer, and Henning Thielemann.
+
+You will need to add:
+
+{\-# LANGUAGE  TemplateHaskell,
+ FlexibleContexts #-\}
+
+to the top of any modules that use this one. 
+-}
+module Data.Lenses.Template (
+   nameDeriveLenses, deriveLenses
+   ) where
+
+import Language.Haskell.TH.Syntax
+  -- (Q, Exp(VarE), Pat(VarP), Dec(ValD), Name(Name), mkOccName, occString, reify, )
+
+import Data.List (nub, )
+import Data.Maybe (catMaybes, )
+import Control.Monad.State
+import Control.Monad (when)
+import Data.Lenses
+
+-- |@deriveLenses n@ where @n@ is the name of a data type
+-- declared with @data@ looks through all the declared fields
+-- of the data type, and for each field ending in an underscore
+-- generates an accessor of the same name without the underscore.
+--
+-- It is "nameDeriveLenses" n f where @f@ satisfies
+--
+-- > f (s ++ "_") = Just s
+-- > f x          = Nothing    -- otherwise
+--
+-- For example, given the data type:
+--
+-- > data Score = Score { p1Score_ :: Int
+-- >                    , p2Score_ :: Int
+-- >                    , rounds   :: Int
+-- >                    }
+--
+-- @deriveLenses@ will generate the following objects:
+--
+-- > p1Score :: (MonadState Score m) => StateT Int m b -> m b
+-- > p1Score = fromGetSet p1Score_ (\x s -> s { p1Score_ = x })
+-- > p2Score :: (MonadState Score m) => StateT Int m b -> m b
+-- > p2Score = fromGetSet p2Score_ (\x s -> s { p2Score_ = x })
+--
+-- It is used with Template Haskell syntax like:
+--
+-- > $( deriveLenses ''TypeName )
+--
+-- And will generate accessors when TypeName was declared
+-- using @data@ or @newtype@.
+deriveLenses :: Name -> Q [Dec]
+deriveLenses n = nameDeriveLenses n stripUnderscore
+
+stripUnderscore :: String -> Maybe String
+stripUnderscore s = do
+    (stem,'_') <- viewR s
+    return stem
+  where
+    -- add to break dependency with Data.List.HT
+    viewR :: [a] -> Maybe ([a], a)
+    viewR [] = Nothing
+    viewR xs = Just (init xs, last xs)
+
+namedFields :: Con -> [VarStrictType]
+namedFields (RecC _ fs) = fs
+namedFields (ForallC _ _ c) = namedFields c
+namedFields _ = []
+
+-- |@nameDeriveLenses n f@ where @n@ is the name of a data type
+-- declared with @data@ and @f@ is a function from names of fields
+-- in that data type to the name of the corresponding accessor. If
+-- @f@ returns @Nothing@, then no accessor is generated for that
+-- field.
+nameDeriveLenses :: Name -> (String -> Maybe String) -> Q [Dec]
+nameDeriveLenses t namer = do
+    info <- reify t
+    reified <- case info of
+                    TyConI dec -> return dec
+                    _ -> fail errmsg
+    (params, cons) <- case reified of
+                 DataD    _ _ params _ cons' _ -> return (params, cons')
+                 NewtypeD _ _ params _ con' _  -> return (params, [con'])
+                 _ -> fail errmsg
+    decs <- makeAccs params . nub $ concatMap namedFields cons
+    when (null decs) $ qReport False nodefmsg
+    return decs
+
+    where
+
+    errmsg = "Cannot derive accessors for name " ++ show t ++ " because"
+          ++ "\n    it is not a type declared with 'data' or 'newtype'"
+          ++ "\n    Did you remember to double-tick the type as in"
+          ++ "\n      $(deriveLenses ''TheType)?"
+
+    nodefmsg = "Warning: No accessors generated from the name " ++ show t
+          ++ "\n    If you are using deriveLenses rather than"
+          ++ "\n    nameDeriveLenses, remember accessors are"
+          ++ "\n    only generated for fields ending with an underscore"
+
+    makeAccs :: [TyVarBndr BndrVis] -> [VarStrictType] -> Q [Dec]
+    makeAccs params vars =
+        fmap (concat . catMaybes) $ mapM (\ (name,_,ftype) -> makeAccFromName name params ftype) vars
+
+    transformName :: Name -> Maybe Name
+    transformName (Name occ f) = do
+        n <- namer (occString occ)
+        return $ Name (mkOccName n) f
+
+    makeAccFromName :: Name -> [TyVarBndr BndrVis] -> Type -> Q (Maybe [Dec])
+    makeAccFromName name params ftype =
+        case transformName name of
+            Nothing -> return Nothing
+            Just n -> fmap Just $ makeAcc name params ftype n
+
+    makeAcc ::Name -> [TyVarBndr BndrVis] -> Type -> Name -> Q [Dec]
+    makeAcc name params ftype accName = do
+        let params' = map (\x -> case x of (PlainTV n _) -> n; (KindedTV n _ _) -> n) params
+        let appliedT = foldl AppT (ConT t) (map VarT params')
+        body <- [|
+                 fromGetSet
+                    ( $( return $ VarE name ) )
+                    ( \x s ->
+                        $( return $ RecUpdE (VarE 's) [(name, VarE 'x)] ) )
+                |]
+        return
+          [ SigD accName (ForallT (map (flip PlainTV SpecifiedSpec) (b:m:params')) [ConT ''MonadState `AppT` appliedT `AppT` (VarT m)] (AppT (AppT ArrowT (AppT (AppT (AppT (ConT ''StateT) ftype) (VarT m)) (VarT b))) (AppT (VarT m) (VarT b))))
+          , ValD (VarP accName) (NormalB body) []
+          ]
+      where
+        b = mkName "b"
+        m = mkName "m"
+
+

